packages/react-components/react-utilities/src/hooks/useMotionPresence.ts

```diff
@@ -0,0 +1,281 @@
+
+  const [shouldRender, setShouldRender] = React.useState(present);
+  const [visible, setVisible] = React.useState(false);
+  const [entering, setEntering] = React.useState(false);
+  const [exiting, setExiting] = React.useState(false);
+  const animating = entering || exiting;
+
+  const [currentElement, setCurrentElement] = React.useState<TElement | null>(null);
+
+  const [setAnimationTimeout, clearAnimationTimeout] = useTimeout();
+
+  const computedStylesRef = React.useRef<CSSStyleDeclaration>({} as CSSStyleDeclaration);
+  const ref: React.RefCallback<TElement> = React.useCallback(node => {
+    if (!node) {
+      return;
+    }
+
+    computedStylesRef.current = getElementComputedStyle(node) as CSSStyleDeclaration;
+    setCurrentElement(node);
+  }, []);
+
+  const notCurrentElement = React.useCallback((target: TElement) => target !== currentElement, [currentElement]);
+
+  const onFinishedMotion = React.useCallback(() => {
+    setEntering(false);
+    setExiting(false);
+  }, []);
+
+  const onStartEntering = React.useCallback(() => {
+    setEntering(true);
+    setVisible(true);
+  }, []);
+
+  const onFinishedEntering = React.useCallback(() => {
+    onFinishedMotion();
+    onEntered();
+  }, [onEntered, onFinishedMotion]);
+
+  const onStartExiting = React.useCallback(() => {
+    setExiting(true);
+  }, []);
+
+  const onFinishedExiting = React.useCallback(() => {
+    onFinishedMotion();
+    setVisible(false);
+    setShouldRender(false);
+    onExited();
+  }, [onExited, onFinishedMotion]);
+
+  const onMotionCanceled = React.useCallback(
+    ({ target }) => {
+      if (notCurrentElement(target)) {
+        return;
+      }
+
+      onFinishedMotion();
+      setVisible(present);
+      setShouldRender(present);
+    },
+    [notCurrentElement, onFinishedMotion, present],
+  );
+
+  React.useEffect(() => {
+    if (present) {
+      setShouldRender(true);
+    }
+  }, [present]);
+
+  React.useEffect(() => {
+    currentElement?.addEventListener('transitioncancel', onMotionCanceled);
+    currentElement?.addEventListener('animationcancel', onMotionCanceled);
+
+    return () => {
+      currentElement?.removeEventListener('transitioncancel', onMotionCanceled);
+      currentElement?.removeEventListener('animationcancel', onMotionCanceled);
+    };
+  }, [currentElement, onMotionCanceled]);
+
+  React.useEffect(() => {
+    if (!currentElement) {
+      return;
+    }
+
+    const { duration, hasTransition, hasAnimation } = getMotionInfo(computedStylesRef.current);
+
+    const animationFrame = requestAnimationFrame(() => {
+      setVisible(present);
+
+      if (!hasTransition && !hasAnimation) {
+        setShouldRender(present);
+      } else {
+        if (present) {
+          onStartEntering();
+        } else {
+          onStartExiting();
+        }
+
+        /**
+         * Use CSS transition duration + 1ms to ensure the animation has finished on both enter and exit states.
```

Comment:
It should be possible to use `transitionend` and `animationend` by tracking the number of CSS properties being animated. Something like:

```tsx
const propCount = numTransitionProperties + numAnimationProperties;
let count = 0;
const handleEnd = (e: TransitionEvent | AnimationEvent) => {
    count++;
    if (count >= propCount) {
         doEnd();
     }
};
```

This could be nice because it would mean less work in computing the duration and give us more accurate events.